keps/sig-instrumentation/1668-trace-popagating/README.md

```diff
@@ -0,0 +1,501 @@
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+This design is inspired by the earlier KEP [Leveraging Distributed Tracing to Understand Kubernetes Object Lifecycles](https://github.com/kubernetes/enhancements/pull/650)
+
+### In-tree changes
+
+#### Trace Utility Package
+
+This package will be able to retrieved span from the span context embedded in the `trace.kubernetes.io/context` object annotation. This package will facilitate propagating traces through kubernetes objects. The exported functions include:
+
+```go
+// WithObject returns a context attached with a Span retrieved from object annotation, it doesn't start a new span
+func WithObject(ctx context.Context, obj meta.Object) (context.Context, error)
+```
+
+When controllers create/update/delete an object A based on another B, we propagate context from B to A. Below is an example to show how deployment propagate trace context  to replicaset.
+
+```diff
+ func (dc *DeploymentController) getNewReplicaSet(d *apps.Deployment, rsList, oldRSs []*apps.ReplicaSet, createIfNotExisted bool) (*apps.ReplicaSet, error) {
++       ctx := httptrace.WithObject(context.Background(), d)
+        existingNewRS := deploymentutil.FindNewReplicaSet(d, rsList)
+@@ -220,7 +227,8 @@ func (dc *DeploymentController) getNewReplicaSet(d *apps.Deployment, rsList, old
+        // hash collisions. If there is any other error, we need to report it in the status of
+        // the Deployment.
+        alreadyExists := false
+-       createdRS, err := dc.client.AppsV1().ReplicaSets(d.Namespace).Create(context.TODO(), &newRS, metav1.CreateOptions{})
++       createdRS, err := dc.client.AppsV1().ReplicaSets(d.Namespace).Create(ctx, &newRS, metav1.CreateOptions{})
+```
+
+#### Add Go context to parameter list
+In OpenTelemetry's Go implementation,  span context is passed down through Go context. This will necessitate the threading of context across more of the Kubernetes codebase, which is a [desired outcome regardless](https://github.com/kubernetes/kubernetes/issues/815). In alpha stage,  we need to change some APIs by adding `ctx context.Context` to parameter list whose parameters doesn't contain context.Context yet. Below APIs will be impacted so far.
+
+| APIs                          | file name                                                    |
+| ----------------------------- | ------------------------------------------------------------ |
+| createPods()                  | pkg/controller/controller_utils.go                           |
+| CreatePodsWithControllerRef() | pkg/controller/controller_utils.go<br />pkg/controller/replication/conversion.go<br />pkg/controller/daemon/daemon_controller.go<br />pkg/controller/replication/conversion.go |
+
+
+### Out-of-tree changes
+
+#### Mutating webhook
```

Comment:
This also needs to address the problem of status updates.  Status updates are writes, but don't indicate a change in desired state.  For example, if we have the following sequence of events

1. Create Deployment
2. Update Deployment.Replicas from 1 -> 2
3. Status update after creation

Since the status update (3) is done with the trace context from (1), it will overwrite the trace context from (2), which I don't think we want.

Reply:
> Status updates are writes, but don't indicate a change in desired state

Sorry, i cannot understand the problem you mentioned. I'm not very much clear about "status update",  may i know which components make the 'status updates' request. I used to think(i hope) the only below 2 events happen.
```
1. Kubectl create deployment
deployment traceID: AAAA

replicaset traceID: AAAA

pod traceID: AAAA

2. Kunectl edit replicaset:1->2
deployment traceID: BBBB

replicaset traceID: BBBB

pod traceID: AAAA
(remain the same because the old pod is not recreated)

pod2 traceID: BBBB
```

I don't know when and how (3)  happens. could you explain more details.

Reply:
I'll try and be more specific--the kube-controller manager does the following:

1. Observe that a deployment is created
2. Create a replicaset based on the deployment
3. Observe the status of the replicaset, when it changes... update the [deployment's status](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.19/#deploymentstatus-v1-apps).  For example, it might update the `availableReplicas` field after a replicaset replica becomes available.

There are a few approaches I can think of that we could take to solve this problem:
1. Don't propagate context (using WithObject) to status update requests.  This means status updates are not linked back to the initial kubectl call.
2. Make the webhook NOT write the annotation if the update is a status update.  For example, we could exclude all calls to subresources using the [resources rule](https://github.com/kubernetes/api/blob/9f65ac4826aa53795a0ef016f5642883129ea4d9/admissionregistration/v1beta1/types.go#L41) in the admission registration config.
3. Propagate [baggage](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/baggage/api.md#baggage-api) in addition to the trace context in annotations.  Add a key in baggage, e.g. `k8s.io/nonmutating: true` to indicate that an API call, like a status update, isn't changing the desired state of an object.  Since baggage is propagated along with the trace context to the webhook, the webhook can skip writing the annotation when baggage contains our key above.
We could provide a helper function to do this, similar to WithObject():
```golang
ctx = WithObject(ctx, obj)
ctx = WithMutating(ctx, false)
client.UpdateFooStatus(ctx, objstatus)
```
We could go even further, and move the WithMutating function inside UpdateFooStatus...

Its probably best to start with option 1 or 2, as they are simpler.

Reply:
Understood. Thank you so much for detailed explaination!
I prefer option 2, as it will bring less in-tree changes.